spmd/tensor/redistribute.py

```diff
@@ -40,38 +45,55 @@ def redistribute_spmd_tensor(
                 partial_spec = cast(_Partial, current)
                 # all_reduce
                 new_local_tensor = device_mesh.all_reduce(
-                    local_tensor, partial_spec.reduce_op
+                    local_tensor, partial_spec.reduce_op, mesh_dim=i
                 )
             else:
-                # for shard, all_gather all shards and return the global tensor
+                assert current.is_shard()
+                # for shard, all_gather all shards and return a tensor that is replicated on the previously sharded dimension
+                shard_spec = cast(Shard, current)
+                new_size = list(local_tensor.size())
+                new_size[shard_spec.dim] *= device_mesh.size(
+                    i
+                )  # only works for evenly sharded tensors
                 new_local_tensor = torch.empty(
-                    input.size(), device=local_tensor.device, dtype=input.dtype
+                    new_size, device=local_tensor.device, dtype=input.dtype
                 )
                 # NOTE: all_gather_base only works well when tensor
                 # sharded on a sequential list of devices
-                device_mesh.all_gather_base(new_local_tensor, local_tensor)
+                device_mesh.all_gather_base(
+                    new_local_tensor,
+                    local_tensor,
+                    mesh_dim=i,
+                    tensor_dim=shard_spec.dim,
+                )
         else:
-            # Case 2: target is Shard
             assert target.is_shard()
+            # Case 2: target is Shard
+
             shard_dim = target.dim  # type: ignore
-            num_chunks = device_mesh.size()
+            num_chunks = device_mesh.size(dim=i)
             assert (
                 input.size(shard_dim) % num_chunks == 0
             ), "Only support chunk sharding evenly now"
-            chunk_size = input.size(shard_dim) // num_chunks
-            my_rank = device_mesh.get_rank()
+            chunk_size = (
+                local_tensor.size(shard_dim) // num_chunks
+            )  # this may already be sharded on a differernt mesh dimension
+            my_rank = device_mesh.get_rank_for_dim(dim=i)
```

Comment:
Is this rank coordinate API only used for checking if the current rank is within the mesh or not? I feel like there might be some other uses, could we document a bit with concrete TODO?

Reply:
It is used in order to calculate offset in the torch.narrow operation below, for replicate -> sharded  redistribution.